pkg/azuredisk/controllerserver.go

```diff
@@ -1219,3 +1241,139 @@ func getEntriesAndNextToken(req *csi.ListSnapshotsRequest, snapshots []compute.S
 
 	return listSnapshotResp, nil
 }
+
+// Creates a new cloud during CreateVolume call to configure the Useragent property of DiskClient
+func SetupNewUserAgentCloud(userAgent string, currentAz *azure.Cloud) (*azure.Cloud, error) {
+	kubeClient := currentAz.KubeClient
+	az := &azure.Cloud{
+		InitSecretConfig: currentAz.InitSecretConfig,
+	}
+	if kubeClient != nil {
+		klog.V(2).Infof("reading cloud config from secret")
+		az.KubeClient = kubeClient
+		if err := InitializeCloudFromSecretWithUserAgent(az, userAgent); err != nil {
+			klog.V(2).Infof("InitializeCloudFromSecret with useragent %s failed with error: %v", userAgent, err)
+		}
+	}
+	if az.TenantID == "" || az.SubscriptionID == "" || az.ResourceGroup == "" {
+		klog.V(2).Infof("could not read cloud config from secret")
+		credFile, ok := os.LookupEnv(DefaultAzureCredentialFileEnv)
+		if ok && strings.TrimSpace(credFile) != "" {
+			klog.V(2).Infof("%s env var set as %v", DefaultAzureCredentialFileEnv, credFile)
+		} else {
+			if util.IsWindowsOS() {
+				credFile = DefaultCredFilePathWindows
+			} else {
+				credFile = DefaultCredFilePathLinux
+			}
+			klog.V(2).Infof("use default %s env var: %v", DefaultAzureCredentialFileEnv, credFile)
+		}
+		var config io.Reader
+
+		config, err := UpdateConfigUserAgent(userAgent, credFile)
+		if err != nil {
+			klog.Errorf("creating azure config from file(%s) with useragent(%s) failed with %v", credFile, userAgent, err)
+			return nil, fmt.Errorf("creating azure config from file(%s) with useragent(%s) failed with %v", credFile, userAgent, err)
+		}
+
+		klog.V(2).Infof("created azure cloud config from file: %s and set useragent: %s successfully", credFile, userAgent)
+		if az, err = azure.NewCloudWithoutFeatureGates(config, false); err != nil {
+			return az, err
+		}
+	}
+	// reassign kubeClient
+	if kubeClient != nil && az.KubeClient == nil {
+		az.KubeClient = kubeClient
+	}
+	return az, nil
+}
+
+// Creates a config buffer that includes a useragent to be used for cloud creation
+func UpdateConfigUserAgent(userAgent string, credFile string) (io.Reader, error) {
```

Comment:
there is a way to reset userAgent easily:
https://github.com/kubernetes-sigs/azuredisk-csi-driver/blob/3bac0081d7046b30e084dc5a56bbb730d0b1f26f/pkg/azuredisk/azure.go#L105-L108

I would suggest reuse `GetCloudProvider` func

Reply:
When I tried utilizing that method, upon inspection of the az object after the InitializeCloudFromConfig function has run, the initialization overwrites it with a nil value since the config does not define it. I can take another look but I'm pretty sure it just gets overwritten